### ISSUE TITLE
negative value for gaussian.rgb

### ISSUE THREAD

**Issue by antonyvan** [joeyan/gaussian_splatting#19]
Hi, 
Thanks you very much for your very clean coding work. 
I am using your code for training some scenes. some of the scenes looks fine, but others just totally fails. I checked the saved final gaussians. gaussian.rgb have negative values, its range is not within 0-1 or 0-255.  I borrowed come code to save it as ply file and tried also convert gaussian.rgb to sperical harmonics dc  by rgb/C0 + 0.5, which doesn't give the correct color when display ply with some open source gaussian splatting viewer.
Here is the code for saving gaussian to ply

import os
from os import makedirs
from plyfile import PlyData, PlyElement

def construct_list_of_attributes(_features_dc, _features_rest, _scaling, _rotation):
    l = ['x', 'y', 'z', 'nx', 'ny', 'nz']
    # All channels except the 3 DC
    for i in range(_features_dc.shape[1]):
        l.append('f_dc_{}'.format(i))
    for i in range(_features_rest.shape[1]):
        l.append('f_rest_{}'.format(i))
    l.append('opacity')
    for i in range(_scaling.shape[1]):
        l.append('scale_{}'.format(i))
    for i in range(_rotation.shape[1]):
        l.append('rot_{}'.format(i))
    return l
    
def save_ply(path, gaussian):
    makedirs(os.path.dirname(path), exist_ok= True)

    xyz = gaussian.xyz.detach().cpu().numpy()
    normals = np.zeros_like(xyz)
    rgb = gaussian.rgb.detach().cpu().numpy()   # * 0.28209479177387814 + 0.5
    sh = gaussian.sh.detach().flatten(start_dim=1).contiguous().cpu().numpy()
    opacities = gaussian.opacity.detach().cpu().numpy()
    scale = gaussian.scale.detach().cpu().numpy()
    rotation = gaussian.quaternion.detach().cpu().numpy()

    dtype_full = [(attribute, 'f4') for attribute in construct_list_of_attributes(rgb,sh, scale, rotation)]
    elements = np.empty(xyz.shape[0], dtype=dtype_full)
    attributes = np.concatenate((xyz, normals, rgb, sh, opacities, scale, rotation), axis=1)
    elements[:] = list(map(tuple, attributes))
    el = PlyElement.describe(elements, 'vertex')
    PlyData([el]).write(path)

The following is one of the results
![iter29999_test_image_8](https://github.com/joeyan/gaussian_splatting/assets/40614822/83d9b947-7e08-4f73-8378-8e74285e3e33)
We got reasonably good result for this scene by official gaussian splatting.
If you want the data to have a test. I could send it you.

**Comment by antonyvan** [comment]
Hi, 
Gaussian splatting viewer works fine with ply files produced by official gaussian splatting code. 

The training with Mip-nerf360 scenes (your provided data) is pretty the same as your results , saved training debug  images looks good,  but gaussian splatting viewer with the saved ply file looks pale.   (Probably its caused by minor difference of official rendering code and yours )

I will check how official version save opacities. (applied sigmoid or not)

**Comment by antonyvan** [comment]
I checked the opacity of official ply files. it doesn't apply sigmoid.
![Screenshot from 2024-04-25 16-28-49](https://github.com/joeyan/gaussian_splatting/assets/40614822/516a6ced-2f44-4f04-90da-f53ca35e5ef0)

**Comment by antonyvan** [comment]
Here is the data
https://drive.google.com/file/d/1k1FsiSPzVD6qR8UYwNEbYrcoY2dmJ8iR/view?usp=sharing
Please download it. once it is done, please let me know I will stop the sharing. 

By the way
I modified your dataloader a little bit so that I don't need create multiple image downsample folders 
![image](https://github.com/joeyan/gaussian_splatting/assets/40614822/b5b9a322-8d51-449c-8735-e3e61e0d1e8e)

**Comment by antonyvan** [comment]
Actually the data I am sharing with you is my custom scene : ship_share. The image above is test result, which looks the training fails. 

the setting I am using to train the ship_share scene is just your default settings . (by the way I turn off the tyro , and directly use SplatConfig, as tyro doesn't allow me the step-by-step run the code.  -- Your code is really clean, I really enjoy it :))

# config = tyro.cli(SplatConfigs,)
config = SplatConfig(
            num_iters=30000,
            adaptive_control_start=1500,
            adaptive_control_end=27500,
            adaptive_control_interval=300,
            reset_opacity_end=27500,
            use_background_end=28000)

**Comment by antonyvan** [comment]
Thanks, I will try these.

**Comment by mokrueger** [comment]
Hi, I have an update on this. @joeyan @antonyvan The issue appears to be the lack of clamping. When accumulating the colors the original authors (or what they say they based their implementation on) disallow negative values for colors by using max(..., 0.0). 

https://github.com/sxyu/svox2/blob/ee80e2c4df8f29a407fda5729a494be94ccf9234/svox2/svox2.py#L777-L780
